aip/general/0164.md

```diff
@@ -0,0 +1,155 @@
+users mistakenly delete resources and need the ability to recover.
+
+If a resource needs to support undelete, the Delete method **must** simply mark
+the resource as having been deleted, but not completely remove it from the
+system. If the method behaves this way, it **should** return the updated
+resource instead of `google.protobuf.Empty`.
+
+Resources that support soft delete **must** have an `expire_time` field as
+described in AIP-148. Additionally, resources **should** include a `DELETED`
+state value if the resource includes a `state` field (AIP-216).
+
+### Undelete
+
+A resource that supports soft delete **should** provide an `Undelete` method:
+
+```proto
+rpc UndeleteBook(UndeleteBookRequest) returns (Book) {
+  post: "/v1/{name=publishers/*/books/*}:undelete"
+  body: "*"
+}
+```
+
+- The HTTP verb **must** be `POST`.
+- The `body` clause **must** be `"*"`.
+- The response message **must** be the resource itself. There is no
+  `UndeleteBookResponse`.
+  - The response **should** include the fully-populated resource unless it is
+    infeasible to do so.
+  - If the create RPC is [long-running](#long-running-undelete), the response
+    message **must** be a `google.longrunning.Operation` which resolves to the
+    resource itself.
+
+### Undelete request message
+
+Undelete methods implement a common request message pattern:
+
+```proto
+message UndeleteBookRequest {
+  // The name of the deleted book.
+  // Format: publishers/{publisher}/books/{book}
+  string name = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference).type = "library.googleapis.com/Book"];
+}
+```
+
+- A `name` field **must** be included. It **should** be called `name`.
+  - The field **should** be annotated as required.
+  - The field **should** identify the [resource type][aip-123] that it
+    references.
+  - The comment for the field **should** document the resource pattern.
+- The request message **must not** contain any other required fields, and
+  **should not** contain other optional fields except those described in this
+  or another AIP.
+
+### Long-running undelete
+
+Some resources take longer to undelete a resource than is reasonable for a
+regular API request. In this situation, the API **should** use a long-running
+operation (AIP-151) instead:
+
+```proto
+rpc UndeleteBook(UndeleteBookRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:undelete"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "Book"
+    metadata_type: "OperationMetadata"
+  }
+}
+```
+
+- The response type **must** be set to the resource (what the return type would
+  be if the RPC was not long-running).
+- Both the `response_type` and `metadata_type` fields **must** be specified.
+
+### List and Get
+
+Soft-deleted resources **should not** show up in `List` (AIP-132) requests by
```

Comment:
requests => responses perhaps? (And maybe "show up" => "be returned"?)

Reply:
Done in both cases.